### ISSUE TITLE
Event formatting agent may be caching regexp results in an event stream

### ISSUE THREAD

**Issue by ohingardail** [cantino/huginn#1222]
The event formatting agent may be failing to reinitialise its (regexp) variables before evaluating a new event in a stream of events.

Example.

Incoming events (A) :
```
[
	{
	  "start_dateTime": "Tuesday 26th January 2016\r\n                7:45 pm - 10:00 pm\r\n                "
	},
	{
	  "start_dateTime": "Thursday 28th January 2016\r\n                8:00 pm - 11:00 pm\r\n            "
	},
	{
	  "start_dateTime": "Friday 29th January 2016\r\n                7:00 pm - 10:30 pm\r\n                "
	}
]
```
Required outgoing events (B):
```
[
	{
	  "start_dateTime": "26th January 2016 7:45 pm",
	  "end_dateTime": "26th January 2016 10:00 pm"
	},
	{
	  "start_dateTime": "28th January 2016 8:00 pm",
	  "end_dateTime": "28th January 2016 11:00 pm"
	},
	{
	  "start_dateTime": "29th January 2016 7:00 pm",
	  "end_dateTime": "29th January 2016 10:30 pm"
	}
]
```
Event formatting agent to convert A to B:
```
{
  "instructions": {
    "start_dateTime": "{{ dateTime.dateStart }} {{ dateTime.timeStart }}",
    "end_dateTime": "{{ dateTime.dateStart }} {{ dateTime.timeEnd }}"
  },
  "mode": "merge",
  "matchers": [
    {
      "path": "{{ start_dateTime  | replace: '\\n', '' )}}",
      "regexp": "\\A\\w+day\\s+(?<dateStart>\\w+\\s+\\w+\\s+\\d{4})\\s+(?<timeStart>\\d{1,2}:\\d{1,2} [ap]m)\\s+-\\s+(?<timeEnd>\\d{1,2}:\\d{1,2}\\s+[ap]m)",
      "to": "dateTime"
    }
  ]
}
```
This agent works if the events are fed through individually (in a dry run or by using the manual event generator for example) but not with a series of incoming events. If this agent is wired up to an other agent that emits a stream of events, then what you get out is :

```
[
	{
	  "start_dateTime": "26th January 2016 7:45 pm",
	  "end_dateTime": "26th January 2016 10:00 pm"
	},
	{
	  "start_dateTime": "26th January 2016 7:45 pm",
	  "end_dateTime": "26th January 2016 10:00 pm"
	},
	{
	  "start_dateTime": "28th January 2016 8:00 pm",
	  "end_dateTime": "28th January 2016 11:00 pm"
	}
]
```
In other words, the event formatting agent appears to be caching the result of the regexp from the previous event if that previous event is not null. The first event in the series is correct, but thereafter each regexp actually evaluates to the result of the previous event's regexp.

The above examples are extracts from a real set of data and agents; it is possible that fields not shown here are germane to the issue.

**Comment by ohingardail** [comment]
I get the same effect if I use the "regexp_replace:" filter instead, or whether I use numerical or verbal indices.

**Comment by knu** [comment]
Ah, there indeed is a bug here.  The matchers method happens to cache the results.  I'll come up with a fix later!

**Comment by knu** [comment]
I think the minimum fix would be add `clear_matches` after the line `interpolation_context.merge(perform_matching(event.payload))`.

**Comment by ohingardail** [comment]
Thank you. I'm relieved it wasn't something stupid I was doing that caused it!

**Comment by knu** [comment]
It turned out caching was a bad idea when receive() could take multiple events at once, so I rewrote the matchers implementation.
Please pull the latest master and give it a try!

**Comment by ohingardail** [comment]
I am now testing the fix - ta.

**Comment by cantino** [comment]
Thanks for fixing it @knu!
